aten/src/ATen/native/cpu/HistogramKernel.cpp

```diff
@@ -0,0 +1,179 @@
+#include <ATen/native/Histogram.h>
+
+#include <ATen/ATen.h>
+#include <ATen/Dispatch.h>
+#include <ATen/Parallel.h>
+
+#include <algorithm>
+#include <mutex>
+#include <tuple>
+
+namespace at { namespace native {
+
+namespace {
+
+constexpr int64_t HISTOGRAM_GRAIN_SIZE = 200;
+
+/* The main algorithm. Maps the elements of input into the bins defined by bin_edges.
+ * Accumulates the total weight in each bin into the hist tensor.
+ */
+enum BIN_SELECTION_ALGORITHM {
+    LINEAR_INTERPOLATION,
+    LINEAR_INTERPOLATION_WITH_LOCAL_SEARCH,
+    BINARY_SEARCH,
+};
+template<typename input_t, BIN_SELECTION_ALGORITHM algorithm>
+void histogram_cpu_contiguous(Tensor& hist, const Tensor& bin_edges,
+        const Tensor& input, const c10::optional<Tensor>& weight) {
+    TORCH_INTERNAL_ASSERT(hist.is_contiguous());
+    TORCH_INTERNAL_ASSERT(bin_edges.is_contiguous());
+    TORCH_INTERNAL_ASSERT(hist.numel() + 1 == bin_edges.numel());
+    TORCH_INTERNAL_ASSERT(input.dim() == 1);
+    TORCH_INTERNAL_ASSERT(!weight.has_value() || weight.value().dim() == 1);
+
+    const int64_t numel_in = input.numel();
+    if (!numel_in) {
+        return;
+    }
+
+    TensorAccessor<input_t, 1> accessor_in = input.accessor<input_t, 1>();
+
+    const auto accessor_wt = weight.has_value()
```

Comment:
Comment here with an elaboration on what this accessor does if weight doesn't have a value